### ISSUE TITLE
[RFC]: Adding JSON to base repo

### ISSUE THREAD

**Issue by rianquinn** [Bareflank/hypervisor#207]
### TL;DR
I would like to add a modified version of the following JSON library to Bareflank:

https://github.com/nlohmann/json

### Description
Specifically, we have a couple of reasons why we need some way to format text. 
+ BFM currently has it's own, custom module file format. This format will need to be modified to something more complicated as we will need a way to specify the default amount of RAM to provide the hypervisor. Ultimately this would mean creating our own version of JSON just to support this. In the future this will only get more complicated as we add support for defining which vCPUs to start, etc... JSON will make easy work of this task
+ The extended APIs will support JSON VMCalls to communicate with the hypervisor. This will also be used by the HyperKernel repo for sending commands for starting / stoping guest VMs including defining things like the amount of RAM, the total number of CPUs and which image file to load. This interface will be similar to [QMP](http://wiki.qemu.org/QMP). Providing base support in Bareflank will provide basic support for this same feature for others, as well as lay the foundation for the Extended APIs and the HyperKernel
+ A file format will be needed for our "[XSM](https://wiki.xenproject.org/wiki/Xen_Security_Modules_:_XSM-FLASK)" wanna-be implementation. JSON is fast, and provides a simple means for defining tree structures that would certainly be needed to support this in the VMM. It's a logical format for denying access to hypercalls. 

The only modification that would be made to the library is the removal of double/float (signed and unsigned integers are already supported). I already tested this approach and it works great (even has compile time checks). 

Also, this library would not be the first library to be added to Bareflank as we already have support for [Hippomocks](http://hippomocks.com/Main_Page) and the [Guidelines Support Library](https://github.com/Microsoft/GSL). 

Just to stress the point... the BFM file format will _have_ to change to something else to support defining more than just filenames for the modules. Since we already plan to support JSON for other things in extension repos, it makes no sense to me, IMO, to use anything other than JSON for the file format for BFM. 

Thoughts?

**Comment by connojd** [comment]
We need a data format that can handle structured types. I think it would be silly create our own when a solution like JSON exists.

**Comment by rianquinn** [comment]
Any one have an issue adding JSON to the bareflank repo based on the above description? If not I'll probably add a PR for this tomorrow.

**Comment by RicoAntonioFelix** [comment]
I agree with using JSON, but what made you gave https://github.com/nlohmann/json preference over the others?

**Comment by rianquinn** [comment]
I like the focus on testing, the fact that it's header only, has native support for signed, unsigned ints, and it was easy to remove double / float support. That being said, I have only looked at a couple. If there is one that you would rather see I'd be more than willing to take a look.

**Comment by rianquinn** [comment]
RapidJSON looks good. It's HUGE. One thing that I like is it has a schema validator which I see people interested in defense liking. It also has as similar level of testing compared to the above. Couple of things I don't like include: it has a LOT of headers that would need to be maintained since we will have to maintain a diff to remove double / float support. It also supports legacy C++ instead of exclusively using modern C++11/14 which means its missing nice features like initializer lists. 

Thoughts?

At the moment the plus for me is the schema code, the minuses are pretty big... not sure if it's worth it. Also... I'll reach out to the developers of the other code to see if schema work is on their radar.

**Comment by RicoAntonioFelix** [comment]
The thought of you needing to do modifications to drop floating points just stand out to me. In that case I presume that the one with the single file is the right choice...

**Comment by rianquinn** [comment]
It only took me about 10mins to make the needed mods to the above JSON library so I know that one was pretty simple. I have a request in to @nlohmann via [Issue #331](https://github.com/nlohmann/json/issues/311). Lets see what they have to say before deciding.

**Comment by rianquinn** [comment]
Based on what @nlohmann said, there seems to be a simple process for adding comment support, and there is an existing schema validator that works with this JSON library, so I think we are good with this one for now unless we can come up any other examples as to why we might want RapidJSON over this.

**Comment by brendank310** [comment]
+1 on adding JSON, though comment support would be nice for some 'self-documentation' in certain instances. Hopefully we can incorporate that change at some point in the future.

**Comment by rianquinn** [comment]
@brendank310 when I make the patch to strip out the double/float support I will see what it takes to get comments working as I agree that would be nice.

**Comment by rianquinn** [comment]
I am in the process of cleaning up the JSON code now. I currently have it working but am attempting to make the following mods:
- Currently the code assumes that if there is no "-", it must be an unsigned number and vice versa. I would like to change this rule to anything that starts with "0x" is an unsigned number and vice versa, and all unsigned numbers are represented in base 16, while all signed numbers are base 10. 
- Going to attempt to add comment support. For now, comments will be represented using "#" since this is not a valid JSON character, I should be able to advance the parser once this is seen. I view this as a stretch goal as the parser is a bit complex. 

Thoughts?

The only downside to this approach is that you will not be able to represent an unsigned integer in base 10, and likewise you will not be able to represent a signed integer in base 16 (not a big deal there). The upside is we will be able to support base 16 numbers, which I feel will occur enough that we should support it. You will still be able to use base 10 numbers, you will just get a signed integer, which you might have to static_cast. Seems like a decent tradeoff to me.

**Comment by rianquinn** [comment]
Ok, after looking at the parsing logic (which uses a hand-rolled regex parser... awesome, but nasty) I think it makes more sense to leave it alone and use strings for hex numbers, and then convert using `std::stoi`. 

Now.. before you complain about that, the number conversion code is basically doing the same thing so the only disadvantage to this approach is the extra memcpy to a `std::string` which is not the end of the world when compared to all of the string parsing that is occurring by using JSON in the first place. In other words, if your code cannot tolerate the extra memcpy, you should not be using JSON, and since we plan to support both, that should work fine. Also note that this only applies to serialization / deserialization (i.e. string <-> number). Simply using JSON has a variant structure is unaffected by any of this. In other words:

```
json j;
j["answer"] = 0x2A
std::cout << j["answer"] == 0x2A << std::endl; // true
```

So, the way this JSON parser works is this:
- If the number starts with "-" you get an integer
- All other valid numbers you get an unsigned integer
- All signed / unsigned numbers use base 10 for conversions
- If you want a hex number, use a string (i.e. `"0x1000"` instead of `0x1000`) and then convert the result yourself.

**Comment by brendank310** [comment]
We need to stop thinking memcpy's are the devil. These are tiny memcpy's for the most part, no sense in jumping through hoops to prevent them. 

I'm a little uneasy about only assuming integers are signed if they are preceded by a '-'. Perhaps we can do something similar to the hex notation for base 10 signed and unsigned ints... "0u400" for unsigned base 10 400. and "0s400" for signed positive base 10 400. We can assume "-400" would indicate signed, but it would probably be cleaner code if we made it uniform as "0s-400". This is bringing back memories of the need for value typing that we had in other json use cases...

**Comment by rianquinn** [comment]
The problem is, the JSON logic assumes the syntax matches a `double` since it still needs to be valid JSON. So, any mod to the format, would require use to modify code that looks like this:

```
basic_json_parser_26:
                    ++m_cursor;
                    {
                        last_token_type = token_type::begin_object;
                        break;
                    }
basic_json_parser_28:
                    ++m_cursor;
                    {
                        last_token_type = token_type::end_object;
                        break;
                    }
basic_json_parser_30:
                    yyaccept = 0;
                    yych = *(m_marker = ++m_cursor);
                    if (yych == 0xBB)
                    {
                        goto basic_json_parser_42;
                    }
                    goto basic_json_parser_5;
basic_json_parser_31:
                    ++m_cursor;
                    if (m_limit <= m_cursor)
                    {
                        fill_line_buffer();
                    }
                    yych = *m_cursor;
basic_json_parser_32:
                    if (yybm[0 + yych] & 128)
                    {
                        goto basic_json_parser_31;
```

That's why I am suggesting we leave the parser the way they have it, and if you need a hex number you use a string. The "-" trick is how that JSON library already works... that's not a mod I would be making.

**Comment by connojd** [comment]
I think we should leave the parsing code unchanged as much as possible.  The less we have to maintain from other codes bases the better.  To me, the extra memcpy is not an issue at all; we're talking about a string that is, in meaningful cases, around 18 characters at most for base 16.

Another thought - stoi and friends handle the minus sign in front. Why not just use those for every number that's parsed?

**Comment by rianquinn** [comment]
Under the hood, that's what it is doing for signed / unsigned its, so by default it's already doing that. Rule of thumb would be if you have a base 10 number, use it as is, otherwise use a string. That should cover 99% of the different use cases we have. Worst case, don't use JSON.

**Comment by nlohmann** [comment]
@rianquinn It seems as info: the code with the `goto`s is generated by re2c. It's way easier to change the original re2c input than the code it generated.

**Comment by rianquinn** [comment]
@nlohmann awesome, yeah I didn't know that. Either way I think I am going to try to keep the code as is for now as the patch was pretty small to get it working, and I agree with @brendank310 and @connojd as I'd prefer to keep the amount we have to maintain to a minimum. Since I didn't change any of the logic, moving to a newer version of your library is as simple as gutting float/double, and your done. 

That being said, if we run into an issue where users really need hex support in the parser, it's nice knowing we won't have to modify that code to make the change.

**Comment by rianquinn** [comment]
Merged
